### PR TITLE
WIP: typescript plugin

### PR DESCRIPTION
**Pull request by stephen**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | yes/
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | no
| Fixed tickets     | #320 
| License           | MIT

This is a work in progress that doesn't cover the entire typescript grammar. I'm currently targeting the [typescript 1.8 syntax](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#A). I've been testing with fixtures locally, but don't have a corpus of test cases checked in yet. In general, this code is pretty rough/prototype stage and there are lots of areas that need cleanup.
 
Some (tentative) design decisions:
- When possible, re-use existing ESTree nodes for the same node
- Otherwise, replicate the official typescript compiler's node
  - Early commits in this PR start off by using making nodes look similar to flow nodes, but I ended up switching away from this as some nodes were hard to adopt because typescript/flow have different syntax or semantics for similar features.
  - Note that the ts compiler uses `.type` in many places for "type annotation" properties, but babel already uses `type` to mean "ast type." In these instances, I've used `.typeAnnotation` in place of `.type`

##### todo
- [ ] add some tests. can we reuse a test suite from somewhere else?
- [ ] decide how the nodes should look/be named
  - maybe follow [typescript-eslint-parser](https://github.com/eslint/typescript-eslint-parser), which prefixes TS-specific nodes with `TS...`
- [ ] complete ts 1.8 grammar parsing
  - [ ] align tsparseproperty name with parsepropertyname?
  - [ ] arrow functions (in particular, `var x = <T>(x:T): T => {};`)
  - [ ] type annotations on classes
  - [ ] ambients
  - [ ] export/import
  - [ ] jsx
- [ ] column locations appear to be off by one?

### CODE REVIEW

**Comment by hzoo** [comment]
Hey @stephen! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217026585)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217026587)

**Comment by hzoo** [comment]
Hey @stephen! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node v4 - linux](https://travis-ci.org/babel/babylon/jobs/217040042)
* [Node v6 - linux](https://travis-ci.org/babel/babylon/jobs/217040044)
* [Node v7 - linux](https://travis-ci.org/babel/babylon/jobs/217040045)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217040046)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217040047)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217040049)

File: src/plugins/typescript.js [hunk 1]
```diff
@@ -0,0 +1,796 @@
...
+const typeIdentifierReservedWords = [
+  "any",
+  "boolean",
+  "number",
+  "string",
+  "symbol",
+  // XXX: note that using `void` in e.g. a TypeParameterList
+  // produces a different error in the ts compiler than the above:
+  // interface X<S, void> {}
+  "void",
```
Comment (DanielRosenwasser) [inline src/plugins/typescript.js]: Don't forget about `object` and `never`!

**Comment by JamesHenry** [comment]
Hey, @stephen! Thanks for mentioning typescript-eslint-parser, that has been over a year's worth of effort and we have covered all kinds of edge cases relating to conversion structure to complement ESTree-powered tooling (i.e. mainly ESLint, but also prettier), and so I definitely thinks it makes sense to align on the AST structure.

@DanielRosenwasser what are your thoughts on this PR in general? It this something that the TS Team is officially onboard with? Or does it fall mainly into the category of a community driven tool, like say, TSLint?

Let me know if I can help!

P.S. How come 1.8? 😄

**Comment by stephen** [comment]
Agree - I think it's a bit unfortunate because it means some nodes will resemble their official TS counterpart, but others will vary greatly.

One example where this is awkward:
```tsx
class A extends B implements C, D {}
```

For the extends syntax, sticking with estree compatibility could mean using `classNode.superClass` and flow's `classNode.superTypeParameters`. However, `implements` isn't in ES, so should we switch to emitting `HeritageClause` there, in the spirit of preserving typescript nodes when possible? (i.e. `classNode.heritageClauses = [ ... ]` - TS compiler uses heritage clauses for both cases.)

I see that [typescript-eslint-parser](https://github.com/eslint/typescript-eslint-parser/blob/master/lib/ast-converter.js#L657) opts for using flow's `ClassImplements` here instead of adding `ExpressionWithTypeParameters` / `HeritageClause`..

It'd also be helpful to know if @DanielRosenwasser / the ts team has any thoughts on estree-compatibility.

re: ts 1.8 grammar, there's no particular reason. I started off targeting 1.8 as a first milestone (and the official grammar spec is fairly complete for it). I'm interested in adding > 1.8 additions after I get <= 1.8 features working.

**Comment by hzoo** [comment]
Hey @stephen! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node v4 - linux](https://travis-ci.org/babel/babylon/jobs/217631957)
* [Node v6 - linux](https://travis-ci.org/babel/babylon/jobs/217631958)
* [Node v7 - linux](https://travis-ci.org/babel/babylon/jobs/217631959)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217631960)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217631961)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217631963)

**Comment by hzoo** [comment]
Hey @stephen! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node v4 - linux](https://travis-ci.org/babel/babylon/jobs/217634832)
* [Node v6 - linux](https://travis-ci.org/babel/babylon/jobs/217634833)
* [Node v7 - linux](https://travis-ci.org/babel/babylon/jobs/217634834)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217634835)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217634836)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/217634838)

**Comment by hzoo** [comment]
Hey @stephen! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node v4 - linux](https://travis-ci.org/babel/babylon/jobs/218323282)
* [Node v6 - linux](https://travis-ci.org/babel/babylon/jobs/218323283)
* [Node v7 - linux](https://travis-ci.org/babel/babylon/jobs/218323284)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/218323285)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/218323286)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/218323288)

**Comment by hzoo** [comment]
Hey @stephen! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/218328783)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/218328786)

**Comment by DanielRosenwasser** [comment]
Hey @JamesHenry, @stephen, @hzoo, sorry for the delay in response. We've also been interested in TypeScript support in Babylon, and one of my teammates @andy-ms has been experimenting with it recently. You can find his work in progress [here](https://github.com/andy-ms/babylon/tree/typescript). From our perspective this is still just an experiment and we want to see how it turns out before taking a hard stance on future support.

I think the work you've done here is pretty awesome, and everyone on the team is pretty impressed with this. That said, it would be great to combine our efforts here. @andy-ms can probably speak better to the diff between his branch and @stephen's branches, but last I checked @andy-ms was pretty far along. I think a good first step would be to take a look at @andy-ms's work and give some feedback.

**Comment by hzoo** [comment]
@stephen Thanks again for this PR! I know it feels bad to had a PR closed but @andy-ms on the TS team has been working on https://github.com/babel/babylon/pull/523 and it's ready for review. I'm going to close this in favor of that PR. Would super appreciate a review if you can spare some time (i'm not that familiar with either flow/ts but can review the babylon core parser changes at least).

**Pull request closed without merging**